src/core/ddsi/src/q_entity.c

```diff
@@ -3894,6 +4209,132 @@ static void add_proxy_builtin_endpoints(
+        if (permissions_hdl != 0) {
+          q_omg_security_register_remote_participant(pp, proxypp, shared_secret, permissions_hdl);
+          match_volatile_secure_endpoints(pp, proxypp);
+        }
+        break;
+
+      case STATE_HANDSHAKE_SEND_TOKENS:
+        DDS_CLOG (DDS_LC_DISCOVERY, &gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") send tokens\n", PGUID (*lpguid), PGUID (*ppguid));
+        q_omg_security_participant_send_tokens(pp, proxypp);
+        break;
+
+      case STATE_HANDSHAKE_OK:
+        DDS_CLOG (DDS_LC_DISCOVERY, &gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") succeeded\n", PGUID (*lpguid), PGUID (*ppguid));
+        update_proxy_participant_endpoint_matching(proxypp, pp);
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+
+      case STATE_HANDSHAKE_TIMED_OUT:
+        DDS_CERROR (&gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") failed: (%d) Timed out\n", PGUID (*lpguid), PGUID (*ppguid), (int)result);
+        if (q_omg_participant_allow_unauthenticated(pp)) {
+          downgrade_to_nonsecure(proxypp);
+          update_proxy_participant_endpoint_matching(proxypp, pp);
+        }
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+      case STATE_HANDSHAKE_FAILED:
+        DDS_CERROR (&gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") failed: (%d) Failed\n", PGUID (*lpguid), PGUID (*ppguid), (int)result);
+        if (q_omg_participant_allow_unauthenticated(pp)) {
+          downgrade_to_nonsecure(proxypp);
+          update_proxy_participant_endpoint_matching(proxypp, pp);
+        }
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+      default:
+        DDS_CERROR (&gv->logconfig, "handshake (lguid="PGUIDFMT" rguid="PGUIDFMT") failed: (%d) Unknown failure\n", PGUID (*lpguid), PGUID (*ppguid), (int)result);
+        ddsi_hsadmin_remove_from_fsm(proxypp->handshake_admin, handshake);
+        break;
+      }
+    }
+  }
+}
+
+static int proxy_participant_check_security_info(struct q_globals *gv, struct proxy_participant *proxypp)
+{
+  int r = 1;
+  struct participant *pp;
+  struct ephash_enum_participant est;
+
+  ephash_enum_participant_init (&est, gv->guid_hash);
+  while (((pp = ephash_enum_participant_next (&est)) != NULL) && (r == 1)) {
+    if (!q_omg_is_similar_participant_security_info(pp, proxypp)) {
+      /* Do not allow secure proxy participant if security info doesn't match. */
+      r = 0;
+    }
+  }
+  ephash_enum_participant_fini(&est);
+  return r;
+}
+
+static int proxy_participant_create_handshake(struct q_globals *gv, struct proxy_participant *proxypp, nn_wctime_t timestamp)
+{
+  int r = 1;
+  struct participant *pp;
+  struct ddsi_handshake *handshake;
+  struct ephash_enum_participant est;
+
+  ddsi_hsadmin_lock(proxypp->handshake_admin);
```

Comment:
Is it really necessary to lock ``proxypp->handshake_admin`` for the duration of the call? It seems like it would make more sense to only lock it for the duration of ``ddsi_hsadmin_register_locked`` (see also ``connect_participant_secure``), in which case it looks like the ``lock``/``unlock`` pair of functions is superfluous.